_posts/2020/2020-02-11-unit-testing-in-react.md

```diff
@@ -0,0 +1,251 @@
+  expect(fetchProducts).toHaveBeenCalledTimes(1)
+  expect(fetchProducts).toHaveBeenCalledWith(1)
+
+  //click the second time
+  fireEvent.click(loadMoreButton)
+  await act(() => Bluebird.delay(500))
+
+  // expect two products
+  expect(getByText('dummy-id-1')).toBeTruthy()
+  expect(fetchProducts).toHaveBeenCalledTimes(2)
+  expect(fetchProducts).toHaveBeenCalledWith(2)
+})
+```
+
+This test can already be taken seriously! It checks whether we called `fetchProducts` with the right arguments and whether we
+use the result of these calls in a meaningful way.
+
+On the repo, this approach is on branch `mocking-fetch-products`.
+
+## Taking it to the next level
+
+The former approach has, nevertheless, its drawbacks. We rely completely on the mock of `fetchProducts`. How can we know if it would
+hit the right API endpoints?
+There's were the awesome `nock` library comes along. You can also mock the HTTP requests! Calling `nock('http://localhost')` will
+mock any requests made to `http://localhost` _inside_ our test! This way we also test the right HTTP requests are being made
+and do not have to mock `fetchProducts` any all!
+
+```javascript
+import React from 'react'
+import { fireEvent, render } from '@testing-library/react'
+import nock from 'nock'
+import ProductList from '../ProductList'
+import Bluebird from 'bluebird'
+import { act } from 'react-dom/test-utils'
+
+it('should render a button to load products', async () => {
+  const scope = nock(`http://localhost`)
+    .get('/products')
+    .query({ page: 1 })
+    .reply(200, {
+      products: [{
+        id: `dummy-id-1`,
+        name: 'dummy-product-1',
+        price: '10 £',
+      }]
+    })
+    .get('/products')
+    .query({ page: 2 })
+    .reply(200, {
+      products: [{
+        id: `dummy-id-2`,
+        name: 'dummy-product-2',
+        price: '20 £',
+      }]
+    })
+
+  const { getByText } = render(<ProductList />)
+
+  const loadMoreButton = getByText('Load more!')
+
+  fireEvent.click(loadMoreButton)
+  await act(() => Bluebird.delay(500))
+
+  // expect one product
+  expect(getByText('dummy-id-1')).toBeTruthy()
+
+  fireEvent.click(loadMoreButton)
+  await act(() => Bluebird.delay(500))
+
+  //expect two products
+  expect(getByText('dummy-id-1')).toBeTruthy()
+  expect(getByText('dummy-id-2')).toBeTruthy()
+
+  scope.done()
+})
+```
+
+You can try this out on the `master` branch of the repo!
+
+Now, we actually test that the endpoint `/products` has been hit with query params `page=1` and `page=2`, thanks to the
+`scope.done()` call, which is an awesome feature of `nock`! It asserts that all of the mocked API endpoint calls have been hit!
+And, furthermore, this tests the component's feature independently of its implementation (so called 'black-box testing').
+It does not matter for the test, for instance, how the `page` parameter is stored in the component. This makes it a lot easier
+when the need for _refactoring_ comes.
```

Comment:
```suggestion
when the component needs to be refactored at a later point.
```

Reply:
done

Reply:
But, it's the same as before. Is this intended?